py-hpc-beeflow: New package

aquan9 opened the pull request:
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

aquan9 committed:
Make an hpc-beeflow package

aquan9 committed:
Merge branch 'spack:develop' into develop

aquan9 committed:
Add neo4j and redis dependencies

aquan9 committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into develop

aquan9 committed:
Add myself as a maintainer

aquan9 commented:
@spackbot help

aquan9 commented:
@spackbot fix style

aquan9 committed:
[@spackbot] updating style on behalf of aquan9

aquan9 committed:
Try to fix styling issues

aquan9 committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into develop

aquan9 committed:
Fix more styling issues.

aquan9 commented:
@spackbot fix style

aquan9 committed:
[@spackbot] updating style on behalf of aquan9

aquan9 commented:
@spackbot request review

tldahlgren reviewed (commented):
I am unable to fetch or stage either version of the package.  Have you built both versions on some platform using spack?

tldahlgren reviewed (commented):
I can confirm the two version sha256s though.

aquan9 commented:
Oops maybe I've set the pypi pointer incorrectly. This is my first time making a Spack package. Let me try to build from this branch locally to try to fix and debug.

bernhardkaindl reviewed (commented):
@aquan9 the version ranges are not correct and it does not concretize:

The version dependencies are listed here:
https://github.com/lanl/BEE/blob/develop/pyproject.toml#L52
```
python = ">=3.8.3,<=3.12.2"
```
You need to translate this to 
```py
depends_on("python@3.8.3:3.12.2", type=("build", "run"))
```
Here are the other dependencies from the file that need to be represented correctly (in case they are not yet):
```py
# Package dependencies
Flask = { version = "^2.0" }
fastapi = { version = "0.109.2" }
uvicorn = { version = "0.27.1" }
neo4j = { version = "^5" }
PyYAML = { version = "^6.0.1" }
flask_restful = "0.3.9"
cwl-utils = "^0.16"
APScheduler = "^3.6.3"
jsonpickle = "^2.2.0"
# Fix for urllib3 2.0 breaking change (similar error to this https://github.com/docker/docker-py/issues/3113)
requests = "<2.29.0"
requests-unixsocket = "^0.3.0"
python-daemon = "^2.3.1"
gunicorn = ">=20.1,<23.0"
# typer version 0.6 and above seem to be throwing an AssertionError with no
# attached info
typer = "^0.5.0"
# Seems to be required for Flux
cffi = "^1.15.1"
celery = { version = "^5.3.4", extras = ["redis", "sqlalchemy"] }
# Fix for poetry/docutils related bug
docutils = "0.18.1"

# Cloud optional dependencies
google-api-python-client = { version = "^2.66.0", optional = true }
python-openstackclient = { version = "^6.0.0", optional = true }
python-heatclient = { version = "^3.1.0", optional = true }
graphviz = "^0.20.3"
networkx = "3.1"
```
I still cannot concretize either version when using the
I converted this to draft to stop the CI.

Maybe I would need to close and reopen it to stop the remaining running GitLab CI task.

bernhardkaindl commented on var/spack/repos/builtin/packages/py-hpc-beeflow/package.py:
```diff
@@ -0,0 +1,32 @@
+    depends_on("neo4j@5.17.0:", type=("run"))
+    depends_on("redis@7.4.0:", type=("run"))
+
+    depends_on("python@3.12.2:", type=("build", "run"))
```
```suggestion
    depends_on("python@3.8.3:3.12.2", type=("build", "run"))
```

bernhardkaindl replied:
This is one of the changes, but you also need to add the other dependencies that I listed in my review comment (from `pyproject.toml`)

aquan9 committed:
fix over-specification of python version

Co-authored-by: Bernhard Kaindl <bernhardkaindl7@gmail.com>

aquan9 commented:
I committed the change for the overspecification of the python version. Thanks for catching that, I started to catch on to that when I was trying to build the package in my local environment and I ran into concretization issues. I suppose I also assumed that Spack was able to figure out the other python dependencies from PyPi. 

I'll start working on adding those other dependencies.

bernhardkaindl commented:
@aquan9:
> [fix over-specification of python version](https://github.com/spack/spack/pull/47025/commits/abd0119eb34b1916cb32a53e67113a1afaec9787)

That's a first step, but you must commit to add all required dependencies mentioned in `pyproject.toml` that I quoted before.

Keep the draft status of the PR until you added all dependencies with the correct version range from pyproject.toml.

One further essential time-saver that you should really use:
- Install the tool `black`, either using your package manager or using `pip` or `pipx` (preferably) and give it to your new package.py file. Amend your commit with the changes from `black`. This will usually fix 99% of all `spack style` errors.
- To be sure, you can run `bin/spack style <path/to/your/package.py` and get a faster feedback about the style check that the GitHub actions will run.

With this, you can save a lot of time that you'd otherwise spend asking @spackbot to fix style (it mostly/only runs `black`!)

aquan9 commented:
Thanks for the tips!

tldahlgren commented:
> * To be sure, you can run `bin/spack style <path/to/your/package.py` and get a faster feedback about the style check that the GitHub actions will run.
> 
> With this, you can save a lot of time that you'd otherwise spend asking @spackbot to fix style (it mostly/only runs `black`!)

You can also use `bin/spack style --fix </path/to/your/package.py>` and spack should correct it for you.  You can optionally provide `--tool <TOOL>` if you want to limit it to `black` or `isort`.

I recently tried this and the command now appears to properly bootstrap the tools it needs.

aquan9 committed:
Add in missing dependencies from pyproject.toml

aquan9 committed:
Fix some typos.

aquan9 commented:
I've translated the requirements from pyproject.toml style to Spack style. Looking through packages.spack.io, it seems like three of the python packages require a version that doesn't yet exist in Spack, and presumably I'll need to go through and make PRs for the inclusion of those versions. 

I put comments on those lines just for now to remind me of which ones I need to add (I know it breaks the style checking).

But also posting it here, in-case others from the BEE team at LANL are tracking this PR. 
Specifically, py-fastapi, py-uvicorn, and py-apscheduler.

bernhardkaindl commented on var/spack/repos/builtin/packages/py-hpc-beeflow/package.py:
```diff
@@ -34,22 +34,22 @@ class PyHpcBeeflow(PythonPackage):
-    depends_on("py-graphviz@0.20.3:", type("run"))
-    depends_on("graphviz", type("run"))
-    depends_on("py-networkx@3.1", type("run"))
+    depends_on("py-uvicorn@0.27.1", type=("run")) #This version doesn't exist yet in Spack
+    depends_on("py-pyyaml@6.0.1:", type=("run"))
+    depends_on("py-flask-restful@0.3.9", type=("run"))
+    depends_on("py-cwl-utils@0.16:", type=("run"))
+    depends_on("py-apscheduler@3.6.3:", type=("run")) #This version doesn't exist yet in Spack
```
@aquan9, two more hints:
- It would be good if you count run `spack install py-hpc-beeflow` on some machine before you push commits.
- If the type is only "run", use `type="run"`.

As these versions do not exist yet in `spack`, I suggest the following:

The these versions are really easy to add to spack, you can add these versions in this PR.

If they are not easy to add, open new PRs that add these versions (one PR for each package).

Note: You are expected to check the `pyproject.toml` files (or other sources) for the version dependencies of those new versions and add those version dependencies as well to these packages.

If you go the route for new PRs, my personal preference would be if you would temporarily close this PR, get the other PRs merged, and then, when all dependencies are merged in to the `develop` branch, create a fresh branch based on develop (or rebase this branch on top of the develop branch with of those merged commits.), and then, when `spack install` of the complete package works and you can run/use the resulting package, open a new PR (or reopen this PR)

aquan9 replied:
Gotcha. I'll go through the pyproject.toml for fastapi@0.109.2 and validate that those dependencies are still correct.

aquan9 replied:
Woah so this sort of sets off a cascade of things that I need to update and validate right? Because beeflow depends on fastapi, but the version isn't there for that, so I need to add it, then validate the dependencies from the pyproject.toml file for fastapi, and I see it needs starlette, but the version of starlette I need doesn't exist, so it's like a continuous rinse and repeat. 

Is the Spack project working on anything to try to automate some of this?

aquan9 replied:
I guess it wasn't so bad, it only took 30 minutes, but it's a process that seems fraught with the potential for user error. 

What I mean is I don't really have a way to validate that the new dependency-dependency packages I've added work correctly besides the use of the dependency in the package that I'm adding.

tldahlgren replied:
Haven't looked at what you've done yet but am glad it wasn't too bad.  Automating it can be problematic too given the amount of variability across different types of projects, repositories, etc.

bernhardkaindl replied:
Yes, there is some work on discovering new versions and updating them on a regular basis,
but to let this fully flourish, more dedicated contributors are likely needed, I think.

I've asked this same question, at least on about if for Python packages, the `pyproject.toml` can be automatically checked for new dependencies of new versions, but the now former Python maintainer (who did an unimaginable amount of work on manually checking dependencies on reviews) said that it can't be done automatically for now as many possible sources for versions (maybe older setup.py/setup.cfg/requriements.txt) should be checked as well.

However, I think we should start somewhere, and that still needs some brainpower to automate getting the e.g. now more and more often used `pyproject.toml` format and extract those versions.

There are some tools that do that, but those that do exist only maintain the latest version of a package, but `spack` needs to support multiple versions of packages, so because of this special capability that spack needs to provides, I have not found an "off-the shelf" solution that could be just used.

The tooling would have to be able to add version dependencies with "when=@<version>:", but sometimes new versions also remove dependencies, so translating all of that will be quite some work.

Of course, baby steps would already help to have checking of missing dependencies for newly submitted additions of newer versions. But such minimal check of missing dependencies would be a good addition to my tool that I write and already use to build (and thereby) check pull requests: https://github.com/spack/build-quality-tools

It could keep the extracted sources and after it built a package, it gets the build location already (using the build log file that is in the same directory) and it could get the pyproject.toml file afte each build and check if it contains packages or versions that are not mentioned in the updated recipe.

Contributions 💯 % welcome! :)

aquan9 replied:
I see... Thank you for the very detailed reply. I'll take a look at that tool you've put together.

aquan9 committed:
Add new fastapi version

aquan9 committed:
Add dependencies for new version of FastAPI

aquan9 committed:
Add new version of starlette needed for new version of fastapi

aquan9 committed:
remove missing dependency note in fastapi

aquan9 committed:
Remove missing fastapi version note

aquan9 committed:
Add new version of py-uvicorn for py-hpc-beeflow

aquan9 committed:
Remove note for missing py-uvicorn version

aquan9 committed:
Add new version of py-apscheduler for py-hpc-beeflow

aquan9 committed:
Fix styling issues

aquan9 committed:
Updating pypi links on all updated packages.

aquan9 commented:
@bernhardkaindl I'm getting some errors now when I'm trying to concretize, and the error isn't super descriptive. It just says: "Cannot select a single "version" for package "py-starlette"", and it says this for a few other packages as well. I'm a little bit confused because I added version 0.36.3, I'm not sure if this error is indicating that it's not able to find the pypi starlette package version that I'm requesting, or if there's some sort of other issue with the packages I'm pulling in. Is there anything in place to validate that other python packages in Spack aren't broken? I mean I'm assuming I'm the one doing something wrong, but I also noticed that there are no maintainers for most of the python package dependencies I added.

aquan9 committed:
Fix wrong version of graphviz dependency

aquan9 committed:
Set upperbound on fastapi dependencies starlette and pydantic

aquan9 commented:
By trial and error I'm learning how the ":" version specification works hahaha.

aquan9 committed:
de-specify the exact version of poetry required for building

aquan9 committed:
Comment out poetry build back-end that I don't think we actually need.

bernhardkaindl commented:
Glad you found out! How the version ranges work is documented on the readthedocs site here:
https://spack.readthedocs.io/en/latest/packaging_guide.html#ranges-versus-specific-versions

Here is a PR that contains an update for the documentation in its diff that may help as well, but the maintainer said it is not needed (duplicate):
https://github.com/spack/spack/pull/35002

aquan9 committed:
py-hpc-beeflow actually installs now

aquan9 commented:
So it looks like, although there is a neo4j package that package doesn't include the neo4j python package (the python drivers) which are in a seperate repo here: https://github.com/neo4j/neo4j-python-driver. So I'll also need to add a package.py for "py-neo4j"

aquan9 committed:
Add new package py-neo4j (python drivers for neo4j)

aquan9 committed:
Add new version of py-tomlkit needed for py-neo4j

aquan9 committed:
Add new version of py-graphviz

aquan9 committed:
Add py-graphviz to py-hpc-beeflow

aquan9 committed:
Set py-typer version to exactly 0.5.0

aquan9 committed:
Fix style issues

aquan9 committed:
Add change to the py-vine package which was not well constrained on python version number

bernhardkaindl reviewed (commented):
Hi @aquan9, If your PR should be able to build your new package completely and is ready for review, you can mark it as "Ready for Review" by clicking the button below the initial CI checks that passed now. Please do so (only) when it all looks finished from your side:

While the PR is marked as draft, GitLab CI is not started on each push of the PR, only GitHub CI runs on draft PRs. But when your PR is ready for review, soonish after setting it ready and on each `git push` to it, GitLab CI is started. The GitLab CI compute resources in the Cloud are sometimes in high demand and don't tell you anything about your new package: It only builds packages that are hard-coded as to be tested. Only changes to those are tested by GitLab CI. Thus, it is mostly final step before merging the PR.

I worked on a script to check PRs by building the changes of the PR, but it has become too time-consuming to continue and spurred some controversy (and most importantly, should be part of CI).

Meanwhile, it is usable and installable as an [extension script on top of the GitHub CLI](https://github.com/spack/gh-spack-pr).

If you are up for it, you can run it when logged in to a machine that can build your PR and has the branch of your PR checked out. I hope the README in the project and see the output of  `gh spack-pr --help` as well.

The GitHub CLI is generally a useful tool for working with PRs (even for only your PRs), and it is easy to install and authorize with the GitHub API.

PS: You can run the script yourself and pass the option `--comment`, which should enable it to build your currently checked-out PR branch and submit the build results as a comment.

PPS: [#pull-requests](https://spackpm.slack.com/archives/C03LVJYGZ7X) is a good way to ask for a review too and a [spack Q&A Telcon is coming up](https://github.com/spack/spack/wiki/Telcon%3A-2024-11-05) The meeting invitation (using Zoom) to join the meeting is the [#general](https://spackpm.slack.com/archives/C5W7NKZJT) Slack Channel of spack: [the invitation is sent each week](https://spackpm.slack.com/archives/C5W7NKZJT/p1730331138729659?thread_ts=1728337005.104389&cid=C5W7NKZJT)

aquan9 commented:
> Hi @aquan9, If your PR should be able to build your new package completely and is ready for review, you can mark it as "Ready for Review" by clicking the button below the initial CI checks that passed now. Please do so (only) when it all looks finished from your side:

Thanks for the advice @bernhardkaindl. I've added a bunch of new versions of packages as I try to resolve the differences in the "concretization" I get from pip installing the package vs. doing a Spack install. 

Currently I'm running into an issue concretizing because my combination of packages isn't able to find a version of Spack's Poetry spec that's able work with all of my other dependencies. 

I'm finding the issue a bit difficult to debug at the moment because of how little information I get when I try to concretize. 
The error I get is just...
```
spack install py-hpc-beeflow
==> Error: failed to concretize `py-hpc-beeflow` for the following reasons:
     1. Cannot select a single "version" for package "py-poetry"
```

So I'm trying to figure out which version of py-poetry I need to add to Spack to make it happy.

tldahlgren commented:
> Can you update the PR to follow the recent change in the repository layout, or close it if not relevant anymore?

You can see the parent directory change in https://github.com/spack/spack/pull/49275.  Basically involves a rebase and `git mv`s to put packages under `$SPACK_ROOT/var/spack/repos/spack_repo/builtin/packages`.
